File: core/exercises/10_dirty_json/solution/python/main.py
"""Exercise 10 — Dirty JSON."""

import json
import os
import re
import urllib.request


def chat_completion(messages: list[dict]) -> str:
    base = os.environ.get("MISTRAL_API_BASE", "https://api.mistral.ai/v1").rstrip("/")
    api_key = os.environ.get("MISTRAL_API_KEY", "")
    url = f"{base}/chat/completions"
    payload = {"model": "mistral-small-latest", "messages": messages}
    body = json.dumps(payload).encode("utf-8")
    request = urllib.request.Request(url, data=body, method="POST")
    request.add_header("Content-Type", "application/json")
    if api_key:
        request.add_header("Authorization", f"Bearer {api_key}")
    with urllib.request.urlopen(request, timeout=30) as response:
        data = json.loads(response.read().decode("utf-8"))
    return data["choices"][0]["message"]["content"]


def extract_json_block(text: str) -> dict:
    match = re.search(r"```(?:json)?\s*(\{.*?\})\s*```", text, re.DOTALL)
    if match:
        return json.loads(match.group(1))
    start = text.find("{")
    end = text.rfind("}")
    if start != -1 and end != -1 and end > start:
        return json.loads(text[start : end + 1])
    raise ValueError("JSON not found in response")


def main() -> None:
    response = chat_completion([{"role": "user", "content": "Return JSON in markdown"}])
    data = extract_json_block(response)
    print(f"PARSED:name={data['name']}")
    print(f"PARSED:score={data['score']}")


if __name__ == "__main__":
    main()

File: core/exercises/10_dirty_json/solution/cpp/main.cpp
#include "codam_llm.hpp"
#include <iostream>
#include <regex>

nlohmann::json extract_json_block(const std::string& text) {
    std::regex fence(R"(```(?:json)?\s*(\{[\s\S]*?\})\s*```)");
    std::smatch match;
    if (std::regex_search(text, match, fence)) {
        return nlohmann::json::parse(match[1].str());
    }
    auto start = text.find('{');
    auto end = text.rfind('}');
    return nlohmann::json::parse(text.substr(start, end - start + 1));
}

int main() {
    nlohmann::json messages = nlohmann::json::array({
        {{"role", "user"}, {"content", "Return JSON in markdown"}},
    });
    std::string response = codam::chat_completion(messages);
    auto data = extract_json_block(response);
    std::cout << "PARSED:name=" << data.at("name").get<std::string>() << "\n";
    std::cout << "PARSED:score=" << data.at("score").get<int>() << "\n";
    return 0;
}
